File: Habitica/src/main/java/com/habitrpg/android/habitica/models/ShopCategory.java
package com.habitrpg.android.habitica.models;

import java.util.List;

public class ShopCategory {

    public String identifier;
    public String text;
    public String notes;
    public Boolean purchaseAll;

    public List<ShopItem> items;

    public String getIdentifier() {
        return identifier;
    }

    public void setIdentifier(String identifier) {
        this.identifier = identifier;
    }

    public String getText() {
        return text;
    }

    public void setText(String text) {
        this.text = text;
    }

    public String getNotes() {
        return notes;
    }

    public void setNotes(String notes) {
        this.notes = notes;
    }

    public Boolean getPurchaseAll() {
        return purchaseAll;
    }

    public void setPurchaseAll(Boolean purchaseAll) {
        this.purchaseAll = purchaseAll;
    }

    public List<ShopItem> getItems() {
        return items;
    }

    public void setItems(List<ShopItem> items) {
        this.items = items;
    }
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/models/shops/ShopCategory.kt
package com.habitrpg.android.habitica.models.shops

import com.google.gson.annotations.SerializedName
import io.realm.RealmList
import java.util.Date

class ShopCategory {
    var identifier: String = ""
    var text: String = ""
    var notes: String = ""
    var path: String = ""
    var purchaseAll: Boolean? = null
    var pinType: String = ""

    @SerializedName("end")
    var endDate: Date? = null

    var itemEndDates: MutableList<Date> = ArrayList()

    var items: MutableList<ShopItem> = ArrayList()

    val endDates: Set<Date>
        get() {
            val dates = itemEndDates.distinct().toMutableSet()
            if (items.isNotEmpty()) {
                dates.addAll(items.mapNotNull { it.endDate })
            }
            endDate?.let { dates.add(it) }
            return dates
        }
}
